target/linux/ath79/dts/qca9563_zte_mf286.dts

```diff
@@ -0,0 +1,253 @@
+				reg = <0x0 0x7840000>;
+				label = "ubi";
+			};
+		};
+	};
+};
+
+&gpio {
+	/* GPIO19 is used as a mask to enable WLAN LED
+	 * in stock firmware, which is controlled directly
+	 * by 5GHz Wi-Fi chip, which currently is inactive
+	 * in OpenWrt */
+	led-wlan {
+		gpio-hog;
+		gpios = <19 GPIO_ACTIVE_LOW>;
+		output-high;
+		line-name = "led:wlan";
+	};
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x080000>;
+				read-only;
+			};
+
+			partition@80000 {
+				label = "u-boot-env";
+				reg = <0x080000 0x020000>;
+				read-only;
+			};
+		};
+	};
+
+	flash@1 {
+		compatible = "spi-nand";
+		reg = <1>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "fota-flag";
+				reg = <0x000000 0x140000>;
+				read-only;
+			};
+
+			caldata: partition@140000 {
+				label = "caldata";
+				reg = <0x140000 0x140000>;
+				read-only;
+
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				cal_art_5000: cal@5000 {
+					reg = <0x5000 0x844>;
+				};
+			};
+
+			mac: partition@280000 {
+				label = "mac";
+				reg = <0x280000 0x140000>;
+				read-only;
+
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				mac_address: mac-address@0 {
+					reg = <0x0 0x6>;
+				};
+			};
+
+			/* This encompasses stock cfg-param, oops, web partitions,
+			 * which can be overwritten safely */
+			ubiconcat0: partition@3c0000 {
+				label = "ubiconcat0";
+				reg = <0x3c0000 0xf40000>;
+			};
+
+			/* Kernel MTD size is increased to 4MB from stock 3MB */
+			partition@1300000 {
+				label = "kernel";
+				reg = <0x1300000 0x400000>;
+			};
+
+			/* This encompasses stock rootfs, data, fota partitions,
+			 * which can be overwritten safely */
+			ubiconcat1: partition@1600000 {
+				label = "ubiconcat1";
+				reg = <0x1700000 0x6900000>;
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0>;
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		phy-mode = "sgmii";
+
+		qca,ar8327-initvals = <
+			0x04 0x00080080 /* PORT0 PAD MODE CTRL */
+			0x7c 0x0000007e /* PORT0_STATUS */
+		>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	nvmem-cells = <&mac_address>;
+	nvmem-cell-names = "mac-address";
+	phy-mode = "sgmii";
+	phy-handle = <&phy0>;
+};
+
+&pcie {
+	status = "okay";
+
+	wifi@0,0,0 {
+		compatible = "qcom,ath10k";
+		reg = <0x0 0 0 0 0>;
+
+		nvmem-cells = <&mac_address>, <&cal_art_5000>;
+		nvmem-cell-names = "mac-address", "calibration";
+		mac-address-increment = <1>;
+	};
+};
+
+&wmac {
+	status = "okay";
+
+	nvmem-cells = <&mac_address>;
+	nvmem-cell-names = "mac-address";
+	mtd-cal-data = <&caldata 0x1000>;
```

Comment:
Starting with 5.16 (openwrt's mac80211 has the backported patches), ath9k can fetch the calibration through nvmem-cells in the same way as ath10k does. If you want, you can give it a whirl by adding:

                cal_caldata_1000: cal@1000 {
                        reg = <0x1000 0x440>;
                };

to the caldata partition definition (in front of the `cal_art_5000`) and extend the nvmem-cells + nvmem-cells-names above to:

      nvmem-cells = <&mac_address>, <&cal_caldata_1000>;
      nvmem-cell-names = "mac-address", "calibration";

Reply:
Sure, will try. I tried that on another device ~2 months ago, but didn't work for me at that time.

Reply:
Ok. Support was added on the 6th November 2021 (this was before ath10k nvmem patch on the 28th).
https://github.com/openwrt/openwrt/commit/d91318662d5c290ce97bfb18c84e4cddc514f33a 

ath10k's caldata loading is easier to check. It prints a "nvmem" during initialization. ath9k wouldn't print much, well except when it fails.

Reply:
For some reason, nvmem caldata didn't work for me on the mentioned MF286A, which uses QCA9888 instead of QCA9880 - it would complain about board files until I added userspace extraction.

Reply:
Oh, QCA9888 could be Wave-2.

From what I remember, there's one QCA9XXX (I think it was QCA9888) that was sold first with a Wave-1 (QCA9887) and later Wave-2 (QCA9886?).

If it's the newer QCA9886-variant of that QCA9888, you need to use the "pre-calibration" in the nvmem-cell-names.

Reply:
Fixed.

Reply:
Regarding QCA9888 - will try later, although I found out that hotplug asked for cal-pci-0000.0.0.0.bin, not pre-cal-pci-*.bin

Reply:
I can't wrap my head about this then. If hotplug asked for cal-pci-.bin, but the ath10k_pci accepted with pre-cal... something is wrong?! Do you have the pciid of that ath10k? It will tell what variant of the QCA9888 it really is.

Reply:
168c:0056, meaning QCA9888, literally. And description on Wikidevi matches output of `iw list`. But hey, as long as it works ;-)

Reply:
That's the QCA9888_2_0_DEVICE_ID. yes "pre-calibration" is definitely the right one.